### PR TITLE
[test][ec2] Add EC2 EFA tests for Training DLCs

### PR DESCRIPTION
**Pull request by saimidu**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
Add EFA Sanity/Integration tests for PyTorch and TensorFlow Training DLCs

### Tests run
Ran new pytest function end-to-end using local system

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [x] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [x] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

### DLC image/dockerfile
pytorch/training/docker/1.13/cu117/Dockerfile.gpu
pytorch/training/docker/1.13/Dockerfile.cpu

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [x] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [x] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [x] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [x] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/testrunner.py [hunk 1]
```diff
@@ -401,6 +402,7 @@ def main():
                 raise Exception(f"EKS cluster {eks_cluster_name} is not in active state")
 
         # Execute dlc_tests pytest command
+        test_path = os.path.join(test_path, "test_efa.py")
```
Comment (saimidu) [inline test/testrunner.py:405]: To be reverted

File: test/dlc_tests/conftest.py [hunk 2]
```diff
@@ -291,15 +305,211 @@ def ec2_instance_role_name(request):
 
 
 @pytest.fixture(scope="function")
-def ec2_instance_ami(request):
-    return request.param if hasattr(request, "param") else UBUNTU_18_BASE_DLAMI_US_WEST_2
+def ec2_instance_ami(request, region):
+    return (
+        request.param
+        if hasattr(request, "param")
+        else UBUNTU_18_BASE_DLAMI_US_EAST_1
```
Comment (jeet4320) [inline test/dlc_tests/conftest.py:312]: We should switch to Ubuntu20 now, maybe in other PR

File: test/dlc_tests/conftest.py [hunk 3]
```diff
@@ -291,15 +305,211 @@ def ec2_instance_role_name(request):
...
+                f"Failed to launch in {availability_zone} due to {e}\n"
+                "Retrying in the next availability zone."
+            )
+            continue
+    if not (response and response["Instances"]):
+        raise RuntimeError(f"Unable to launch {ec2_instance_type} instances in {region}")
+
+    instances = response["Instances"]
+
+    def terminate_efa_instances():
```
Comment (jeet4320) [inline test/dlc_tests/conftest.py:396]: do we delete key pairs as well

File: test/dlc_tests/conftest.py [hunk 4]
```diff
@@ -291,15 +305,211 @@ def ec2_instance_role_name(request):
...
+            ec2_utils.create_name_tags_for_instance(
+                worker_instance_id, f"{instance_name_prefix}_worker_{i}", region
+            )
+            ec2_utils.check_instance_state(worker_instance_id, state="running", region=region)
+            ec2_utils.check_system_state(
+                worker_instance_id, system_status="ok", instance_status="ok", region=region
+            )
+            print(f"Worker instance {worker_instance_id} is ready")
+
+    if ec2_instance_type in ["p4d.24xlarge", "p4de.24xlarge", "trn1.32xlarge"]:
```
Comment (jeet4320) [inline test/dlc_tests/conftest.py:425]: if we are using same config for trn1, then trn1n is also launched

File: test/dlc_tests/container_tests/bin/efa/efa_test.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
```
Comment (jeet4320) [inline test/dlc_tests/container_tests/bin/efa/efa_test.sh:1]: this test gets added automatically by EFA at `/opt/amazon/efa/test/efa_test.sh`

File: test/dlc_tests/container_tests/bin/efa/testEFA [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    open_mpi_lib_path=/opt/amazon/openmpi/lib
+else
+    open_mpi_lib_path=/opt/amazon/openmpi/lib64
+fi
+
+check_ring_single_node(){
+    echo "Running ring"
+    $openmpi_path_for_efa -n 3 --host localhost --oversubscribe \
+        -x RDMAV_FORK_SAFE=1 -x FI_EFA_USE_DEVICE_RDMA=1 -x NCCL_PROTO=simple --mca pml ^cm \
+        -x LD_LIBRARY_PATH=/usr/local/cuda/efa/lib:/usr/local/cuda/lib:/usr/local/cuda/lib64:/usr/local/cuda:$LD_LIBRARY_PATH \
```
Comment (jeet4320) [inline test/dlc_tests/container_tests/bin/efa/testEFA:32]: can you check if there is dir `/usr/local/efa` as I see DLC install ofi plugin at `/usr/local` https://github.com/aws/deep-learning-containers/blob/master/pytorch/training/docker/1.13/py3/cu117/Dockerfile.gpu#LL292C35-L292C45

File: test/dlc_tests/container_tests/bin/efa/testEFA [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        else
+            echo "***************************** check_efa_nccl_all_reduce failed for cuda version ${CUDA_VERSION} *****************************"
+        fi
+        check_efa_nccl_all_reduce_performance $CUDA_VERSION
+    done
+}
+
+check_ring_single_node
+check_mutlinode_nccl_transfer
+#check_efa_nccl_all_reduce
```
Comment (jeet4320) [inline test/dlc_tests/container_tests/bin/efa/testEFA:101]: is comment expected

File: test/dlc_tests/conftest.py (hunk 2 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:312]: Agreed. Let's do this in a separate PR.

File: test/dlc_tests/conftest.py (hunk 3 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:396]: Yes, refer to line 355 in this file, where we add a finalizer to delete the keypair.

File: test/dlc_tests/conftest.py (hunk 4 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:425]: Replacing this with the following:
```
    num_efa_interfaces = ec2_utils.get_num_efa_interfaces_for_instance_type(
        ec2_instance_type, region=region
    )
    if num_efa_interfaces and num_efa_interfaces > 1:
```
Is this good as a future-proof check for whether elastic ip allocation is needed?

File: test/dlc_tests/container_tests/bin/efa/efa_test.sh (hunk 5 above)
Comment (saimidu) [reply test/dlc_tests/container_tests/bin/efa/efa_test.sh:1]: Summarizing offline discussion here - This script exists from EFA installation on DLAMIs, but doesn't get automatically created in DLC EFA installation. Therefore adding this script to the test scripts is necessary.

File: test/dlc_tests/container_tests/bin/efa/testEFA (hunk 6 above)
Comment (saimidu) [reply test/dlc_tests/container_tests/bin/efa/testEFA:32]: No, the EFA libs come from `/opt/amazon/efa/lib`. There doesn't seem to be a `/usr/local/efa`.

File: test/dlc_tests/container_tests/bin/efa/testEFASanity [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+fi_info -p efa
+fi_info -p efa -t FI_EP_RDM | grep 'FI_EP_RDM'
+
+apt-get update && apt-get install -y kmod
+
+# check if ib_uverbs is present
+lsmod | grep ib_uverbs
+
+# ensure that the security group created is configured correctly
+/test/bin/efa/efa_test.sh
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/dlc_tests/container_tests/bin/efa/testEFASanity:21]: The `testEFASanity` script and the `efa_test.sh` script seem to be testing different things. Are the tests run by one of the scripts redundant?

File: test/dlc_tests/ec2/test_efa.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+@pytest.mark.model("N/A")
+@pytest.mark.integration("efa")
+@pytest.mark.usefixtures("sagemaker")
+@pytest.mark.multinode(2)
+@pytest.mark.parametrize("ec2_instance_type,region", EC2_EFA_GPU_INSTANCE_TYPE_AND_REGION)
+@pytest.mark.skipif(
+    is_pr_context() and not is_efa_dedicated(),
+    reason="Skip EFA test in PR context unless explicitly enabled",
+)
+def test_efa_tensorflow(
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/dlc_tests/ec2/test_efa.py:68]: Summarizing offline discussion here - The use of `pytorch_training` allows tests to parametrize tests specifically for PyTorch DLCs, and `tensorflow_training` allows only TensorFlow DLCs.

File: test/test_utils/ec2.py [hunk 10]
```diff
@@ -13,18 +13,20 @@
 from invoke import run
 from packaging.version import Version
 from packaging.specifiers import SpecifierSet
-from tenacity import retry, stop_after_attempt, wait_fixed
+from tenacity import (
+    retry, stop_after_attempt, stop_after_delay, wait_fixed, wait_random_exponential
+)
```
Comment (saimidu) [inline test/test_utils/ec2.py:18]: Formatting check failed here. Will fix in next revision of the PR alongside any other changes requested.

**Review by jeet4320 — APPROVED** [review]

**Comment by saimidu** [comment]
All tests pass on PT 1.13 and TF 2.12 Training GPU DLCs at commit [0019067](https://github.com/aws/deep-learning-containers/pull/2970/commits/001906762c9d25e84335c6c6c268f3ea0e4b43db).

**Comment by saimidu** [comment]
Test passes on [8a847b1](https://github.com/aws/deep-learning-containers/pull/2970/commits/8a847b136a8eee228fd3d81009190e19c3533930) for PT 2.0 and TF 2.12 DLCs.

Test passes on [a29076d](https://github.com/aws/deep-learning-containers/pull/2970/commits/a29076d53b4974c6bf5f221c395e94cf9ed9774f) for PT 1.13 DLCs.

File: test/test_utils/__init__.py [hunk 11]
```diff
@@ -35,6 +35,8 @@
 DEFAULT_REGION = "us-west-2"
 # Constant to represent region where p3dn tests can be run
 P3DN_REGION = "us-east-1"
+# Constant to represent region where p4de tests can be run
+P4DE_REGION = "us-east-1"
```
Comment (tejaschumbalkar) [inline test/test_utils/__init__.py:39]: is this temporary?

File: test/dlc_tests/container_tests/bin/efa/testEFA [hunk 12]
```diff
@@ -0,0 +1,105 @@
+#!/bin/bash
```
Comment (tejaschumbalkar) [inline test/dlc_tests/container_tests/bin/efa/testEFA:1]: whats the source for this script?

File: test/dlc_tests/conftest.py [hunk 13]
```diff
@@ -908,6 +1122,9 @@ def pytest_configure(config):
         "markers", "processor(cpu/gpu/eia/hpu): explicitly mark which processor is used"
     )
     config.addinivalue_line("markers", "efa(): explicitly mark to run efa tests")
+    config.addinivalue_line(
+        "markers", "allow_p4de_use(): explicitly mark to allow test to use p4de instance types"
```
Comment (tejaschumbalkar) [inline test/dlc_tests/conftest.py:1126]: why is this needed? is it to conserve on p4de usage?

File: test/dlc_tests/ec2/test_efa.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+@pytest.mark.integration("efa")
+@pytest.mark.usefixtures("sagemaker")
+@pytest.mark.allow_p4de_use
+@pytest.mark.multinode(2)
+@pytest.mark.parametrize("ec2_instance_type,region", EC2_EFA_GPU_INSTANCE_TYPE_AND_REGION)
+@pytest.mark.skipif(
+    is_pr_context() and not is_efa_dedicated(),
+    reason="Skip EFA test in PR context unless explicitly enabled",
+)
+def test_efa_pytorch(
```
Comment (tejaschumbalkar) [inline test/dlc_tests/ec2/test_efa.py:45]: lets add description on what the test does. I see two test scripts are run.

File: test/dlc_tests/ec2/test_efa.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+            WORKER_CONTAINER_NAME,
+            worker_connection,
+            BUILD_ALL_REDUCE_PERF_CMD,
+            timeout=180,
+            asynchronous=True,
+        )
+        build_all_reduce_perf_promises.append(promise)
+
+    # Configure master node SSH client-side configurations
+    _setup_master_efa_ssh_config(master_connection)
```
Comment (tejaschumbalkar) [inline test/dlc_tests/ec2/test_efa.py:135]: nit: efa/mpi setup can be in a bash file, that can directly be executed on the container.

File: test/dlc_tests/conftest.py [hunk 16]
```diff
@@ -241,40 +241,61 @@ def ec2_public_ip(request):
     return request.param
 
 
-@pytest.fixture(scope="session")
-def region():
-    return os.getenv("AWS_REGION", DEFAULT_REGION)
+@pytest.fixture(scope="function")
```
Comment (tejaschumbalkar) [inline test/dlc_tests/conftest.py:244]: why are we switching to using `function` in all the fixtures?

File: test/test_utils/__init__.py [hunk 17]
```diff
@@ -35,6 +35,8 @@
 DEFAULT_REGION = "us-west-2"
```
Comment (tejaschumbalkar) [inline test/test_utils/__init__.py:35]: We are using explicit regions declared here for parameterizing `"ec2_instance_type,region"` in the test.
However, the same is not used in the `region` declaration in conftest

File: test/dlc_tests/conftest.py [hunk 18]
```diff
@@ -241,40 +241,61 @@ def ec2_public_ip(request):
...
+    Parametrize with a reduced list of availability zones for particular instance types for which
+    capacity has been reserved in that AZ. For other instance types, parametrize with list of all
+    AZs in the region.
+    :param ec2_client: boto3 Client for EC2
+    :param ec2_instance_type: str instance type for which AZs must be determined
+    :param region: str region in which instance must be created
+    :return: list of str AZ names
+    """
+    allowed_availability_zones = None
+    if ec2_instance_type in ["p4de.24xlarge"]:
```
Comment (tejaschumbalkar) [inline test/dlc_tests/conftest.py:261]: isnt the region to instance type configuration already validated?

File: test/dlc_tests/conftest.py (hunk 16 above)
Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:244]: for all usecases, even though just efa needs it? it might increase the boto3 calls and lead to throttling.

File: test/dlc_tests/conftest.py [hunk 19]
```diff
@@ -291,15 +312,207 @@ def ec2_instance_role_name(request):
...
+    def delete_ssh_keypair():
+        if test_utils.is_pr_context():
+            test_utils.destroy_ssh_keypair(ec2_client, key_filename)
+        else:
+            with open(KEYS_TO_DESTROY_FILE, "a") as destroy_keys:
+                destroy_keys.write(f"{key_filename}\n")
+
+    request.addfinalizer(delete_ssh_keypair)
+
+    instance_name_prefix = f"CI-CD {ec2_key_name}"
```
Comment (tejaschumbalkar) [inline test/dlc_tests/conftest.py:357]: nit: we can add `EFA` to the name be be explicit

File: test/dlc_tests/conftest.py [hunk 20]
```diff
@@ -291,15 +312,207 @@ def ec2_instance_role_name(request):
...
 
 
 @pytest.fixture(scope="function")
 def ei_accelerator_type(request):
     return request.param if hasattr(request, "param") else None
 
 
+@pytest.mark.timeout(300)
+@pytest.fixture(scope="function")
+def efa_ec2_instances(
```
Comment (tejaschumbalkar) [inline test/dlc_tests/conftest.py:332]: if there is same common code between this and `ec2_instances`, can we define common code and reuse it in both the methods?

File: pytorch/training/docker/1.13/py3/cu117/Dockerfile.gpu [hunk 21]
```diff
@@ -155,7 +155,7 @@ RUN mkdir /tmp/openmpi \
  && cd / \
  && rm -rf /tmp/openmpi
 
-ENV PATH="$OPEN_MPI_PATH/bin:$PATH"
+ENV PATH="$OPEN_MPI_PATH/bin:$EFA_PATH/bin:$PATH"
```
[the remark this answers is not in the document]
Comment (saimidu) [reply pytorch/training/docker/1.13/py3/cu117/Dockerfile.gpu:158]: Yes, we did, but the EFA_PATH doesn't contain any binaries that would be used for actual customer jobs. It only contains basic utility binaries that can still be used by passing the full path to it, like `/opt/amazon/efa/bin/fi_info` instead of just `fi_info`.

File: test/test_utils/__init__.py (hunk 11 above)
Comment (saimidu) [reply test/test_utils/__init__.py:39]: No, this is a permanent code-change, which can later be modifed if we decide to change the region in which we launch p4de instances.

File: test/dlc_tests/container_tests/bin/efa/testEFA (hunk 12 above)
Comment (saimidu) [reply test/dlc_tests/container_tests/bin/efa/testEFA:1]: This is taken from DLAMI integration tests for EFA.

**Review by tejaschumbalkar — COMMENTED** [review]
added comments

File: test/dlc_tests/conftest.py (hunk 13 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:1126]: That's right. If a test does not have `allow_p4de_use` as one of its markers, it will be skipped when the instance type environment variable on the test job is defined as `p4de.24xlarge`.

File: test/dlc_tests/ec2/test_efa.py (hunk 15 above)
Comment (saimidu) [reply test/dlc_tests/ec2/test_efa.py:135]: Made the decision to keep the SSH/MPI-Hosts setup in the python code so that it can be found and modified more easily outside of the actual test scripts, and to make the test setup more "explainable" and less of a black-box.

File: test/dlc_tests/conftest.py (hunk 16 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:244]: By making the scope for the `region` fixture `function` instead of `session`, the region can be assigned based on the instance type used for the test, rather than being hardcoded and badly-reassigned like we currently do for p3dn instances. This prevents other fixtures that also need `ec2_client/resource/other regional clients` from erroring-out (or failing silently).

But now, since `region` is a function-scoped fixture, other fixtures that depend on `region` also need to be function-scoped, because a session-scoped fixture cannot be defined based on a function-scoped fixture.

File: test/dlc_tests/conftest.py (hunk 18 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:261]: We only want to use hardcoded availability-zone lists if the instance type being launched and the region both match pre-existing capacity reservations.

While the helper function to assign `instance-type + region` always associates only the region with capacity-reservations to the instance-type for now, we might have changes in the future where we may try to launch instances in other regions.

That's why we use the hardcoded AZ lists in specific scenarios, and get AZ IDs using the helper function otherwise.

File: test/dlc_tests/conftest.py (hunk 19 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:357]: The test name would have EFA in it already. Also, the instances launched would be renamed with `master` and `worker` suffixes. Is there any other reason we might need `efa` in the name?

File: test/dlc_tests/conftest.py (hunk 20 above)
Comment (saimidu) [reply test/dlc_tests/conftest.py:332]: For now, there are some differences in the way the instance is launched for `ec2_instance` and `efa_ec2_instances`, such as a helper function to retry instance-launch, etc.

I'd like for us to have a separate PR to consolidate both functions, by using that retry+backoff functionality on instance launch for all tests, because this PR has already gotten complicated as-is.

In case the retry-backoff functionality is bad/flaky, the only tests affected right now will be the EFA tests, as opposed to all EC2 tests.

Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:332]: ack

File: test/dlc_tests/conftest.py (hunk 19 above)
Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:357]: if efa is already in the test name, that would work

File: test/dlc_tests/conftest.py (hunk 16 above)
Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:244]: lets observe if we see throttling for boto calls in the CI

File: test/dlc_tests/conftest.py (hunk 18 above)
Comment (tejaschumbalkar) [reply test/dlc_tests/conftest.py:261]: ack

**Review by tejaschumbalkar — APPROVED** [review]
